src/ZkFold/Symbolic/Data/UInt.hs

```diff
@@ -127,522 +102,417 @@ cast n =
-            solve = do
-                is <- runCircuit x
-                js <- runCircuit y
-                case V.fromVector $ Z.zip is js of
-                  []              -> return []
-                  [(i, j)]        -> solve1 i j
+            t :: BaseField c
+            t = (one + one) ^ registerSize @(BaseField c) @n @r - one
+
+            solve :: MonadCircuit i (BaseField c) m => Vector (NumberOfRegisters (BaseField c) n r) i -> Vector (NumberOfRegisters (BaseField c) n r) i -> m (Vector (NumberOfRegisters (BaseField c) n r) i)
+            solve xv yv = do
+                case V.fromVector $ Z.zip xv yv of
+                  []              -> return $ V.unsafeToVector []
+                  [(i, j)]        -> V.unsafeToVector <$> solve1 i j
                   ((i, j) : rest) -> let (z, w) = Haskell.last rest
                                          (ris, rjs) = Haskell.unzip $ Haskell.init rest
-                                      in solveN (i, j) (ris, rjs) (z, w)
+                                      in V.unsafeToVector <$> solveN (i, j) (ris, rjs) (z, w)
 
-            solve1 :: MonadBlueprint i a m => i -> i -> m [i]
+            solve1 :: MonadCircuit i (BaseField c) m => i -> i -> m [i]
             solve1 i j = do
                 z <- newAssigned (\v -> v i - v j)
-                _ <- expansion (highRegisterSize @a @n @r) z
+                _ <- expansion (highRegisterSize @(BaseField c) @n @r) z
                 return [z]
 
-            solveN :: MonadBlueprint i a m => (i, i) -> ([i], [i]) -> (i, i) -> m [i]
+            solveN :: MonadCircuit i (BaseField c) m => (i, i) -> ([i], [i]) -> (i, i) -> m [i]
             solveN (i, j) (is, js) (i', j') = do
                 s <- newAssigned (\v -> v i - v j + fromConstant (t + one))
-                (k, b0) <- splitExpansion (registerSize @a @n @r) 1 s
+                (k, b0) <- splitExpansion (registerSize @(BaseField c) @n @r) 1 s
                 (zs, b) <- flip runStateT b0 $ traverse StateT (Haskell.zipWith fullSub is js)
                 k' <- newAssigned (\v -> v i' - v j')
                 s' <- newAssigned (\v -> v k' + v b - one)
-                _ <- expansion (highRegisterSize @a @n @r) s'
+                _ <- expansion (highRegisterSize @(BaseField c) @n @r) s'
                 return (k : zs <> [s'])
 
 
-            fullSub :: MonadBlueprint i a m => i -> i -> i -> m (i, i)
+            fullSub :: MonadCircuit i (BaseField c) m => i -> i -> i -> m (i, i)
             fullSub xk yk b = do
                 k <- newAssigned (\v -> v xk - v yk)
                 s <- newAssigned (\v -> v k + v b + fromConstant t)
-                splitExpansion (registerSize @a @n @r) 1 s
+                splitExpansion (registerSize @(BaseField c) @n @r) 1 s
 
-    strictMul (UInt x) (UInt y) = UInt (circuitF $ V.unsafeToVector <$> solve)
+    strictMul (UInt x) (UInt y) = UInt $ symbolic2F x y (\u v -> V.unsafeToVector $ V.fromVector u * V.fromVector v) solve
```

Comment:
Pointwise list addition doesn't do what is needed here, I would write a helper function `Arithmetic a => Vector (NumberOfBits ...) a -> Natural` which does the same what `ToConstant (UInt n r (Interpreter a)) Natural` does